Improve details for user error events in OIDC protocol endpoints

thomasdarimont opened the pull request:
Fixes #29166

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

thomasdarimont committed:
Improve details for user error events in protocol endpoints (#29166)

Fixes #29166

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java:
```diff
@@ -244,7 +249,9 @@ protected String getRequestedScopes() {
         if (!validScopes) {
             event.error(Errors.INVALID_REQUEST);
-            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_SCOPE, "Invalid scopes: " + scope,
+            String errorMessage = "Invalid scopes: " + scope;
```
Same concern like above (the fact that we're returning scopes inside the CorsErrorResponseException is likely also far from ideal, so if you can change in the PR when on this, it can be nice :-)

thomasdarimont replied:
Same as above. I'll remove the received scopes from the returned response. How about logging a warning here with the invalid scope?

mposolda replied:
Thanks. Also sounds good to me to log the warning. Sending back the valid scopes is also a good idea.

thomasdarimont replied:
I updated the code, the error response now looks like this:
```bash
curl -s \
  -d "grant_type=password" \
  -d "username=admin" \
  -d "password=admin" \
  -d "client_id=admin-cli" \
  -d "scope=unknown" \
  http://localhost:8081/auth/realms/master/protocol/openid-connect/token 
```

```json
{"error":"invalid_scope","error_description":"Invalid scopes. Scopes must be contained in [address, email, microprofile-jwt, offline_access, organization, phone, profile]"}
```

mposolda reviewed (commented):
@thomasdarimont Thanks, the additional details in the event are useful!

Added some minor comments inline, but I think that this also causes some test failures `AccessTokenTest`, `OAuth2DeviceAuthorizationGrantTest`, `OAuthProofKeyForCodeExchangeTest`, `OID4VCIssuerEndpointTest` (not sure about last one as it is flaky in general and it is failing in some other PRs as well...)

thomasdarimont commented on services/src/main/java/org/keycloak/protocol/oidc/endpoints/AuthorizationEndpoint.java:
```diff
@@ -313,8 +316,11 @@ private void updateAuthenticationSession() {
                 boolean essential = Boolean.parseBoolean(authenticationSession.getClientNote(Constants.FORCE_LEVEL_OF_AUTHENTICATION));
                 if (essential) {
-                    logger.errorf("Requested essential acr value '%s' is not a number and it is not mapped in the ACR-To-Loa mappings of realm or client. Please doublecheck ACR-to-LOA mapping or correct ACR passed in the 'claims' parameter.", acr);
+                    String errorMessage = "Requested essential acr value '%s' is not a number and it is not mapped in the ACR-To-Loa mappings of realm or client.";
```
Good point. I'll adapt the `reason` value as you suggested.

mposolda replied:
Thanks

thomasdarimont committed:
Adapt test to check for proper error type in event

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Adapt test to check for more detailed error response.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Unified error messages for PKCE error responses

- Adapted tests to reflect the change

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Add debug logging with actual invalid scope

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
@mposolda I revised the code as discussed and fixed the test failures, which were mostly checking for the errorDescription in the response. All tests pass now locally including `OID4VCTest` which I didn't need to adapt.

thomasdarimont commented on services/src/main/java/org/keycloak/protocol/oidc/grants/AuthorizationCodeGrantType.java:
```diff
@@ -127,24 +130,32 @@ public Response process(Context context) {
         if (redirectUri != null && !redirectUri.equals(redirectUriParam)) {
-            event.error(Errors.INVALID_CODE);
-            logger.tracef("Parameter 'redirect_uri' did not match originally saved redirect URI used in initial OIDC request. Saved redirectUri: %s, redirectUri parameter: %s", redirectUri, redirectUriParam);
+            event.error(Errors.INVALID_REDIRECT_URI);
```
Here I used the more appropriate `Errors.INVALID_REDIRECT_URI` instead of `Errors.INVALID_CODE`

thomasdarimont commented on services/src/main/java/org/keycloak/protocol/oidc/grants/AuthorizationCodeGrantType.java:
```diff
@@ -127,24 +130,32 @@ public Response process(Context context) {
         if (!client.getClientId().equals(clientSession.getClient().getClientId())) {
-            event.error(Errors.INVALID_CODE);
-            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_GRANT, "Auth error", Response.Status.BAD_REQUEST);
+            event.error(Errors.INVALID_CLIENT);
```
Here I used the more appropriate `Errors.INVALID_CLIENT` instead of `Errors.INVALID_CODE`

thomasdarimont committed:
Cleanup

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented on services/src/main/java/org/keycloak/protocol/oidc/grants/OAuth2GrantTypeBase.java:
```diff
@@ -243,8 +254,15 @@ protected String getRequestedScopes() {
+            logger.debugf("Invalid scopes: %s. realm=%s client_id=%s", scope, realm.getName(), client.getClientId());
             event.error(Errors.INVALID_REQUEST);
-            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_SCOPE, "Invalid scopes: " + scope,
+            Set<String> allowedScopes = Stream.concat( //
```
Here we collect all allowed scope names that can be requested (and represented in the token).

ahus1 reviewed (changes requested):
Hi @thomasdarimont - thank you for improving the logging and the events. 

Something I learned just recently: You'll need to call `event.detail()` before calling `event.error()`, as the `event.error()` will send out and store the event. If you do it the other way around, the details will get lost. 

About the available scopes: I agree to store this information in the event, which will then be logged and depending on the configuration also stored in the database. 

I then disagree to share the whole list of scopes with the caller, as I would consider this an information leak if this is shared with an unauthorized caller aka public client (CWE-200). Maybe we could have different messages for the event and returned to the caller.

I also see that you're adding `log.warn()` in some places. As the default setting is to log error events, this is IMHO redundant. Maybe you want to reconsider some of those log messages.

thomasdarimont committed:
Adapt code according to code review

- Adapted code to call event.detail(..) before event.error(..)
- Adapted remaining OIDC endpoints
- Reverted log levels to previous values in OAuth2GrantTypeBase

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Cleanup

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
Thanks @ahus1 for taking a look and your helpful suggestions. I totally missed the subtle relationship between `event.detail(..)`  and `event.error(..)`. 

Regarding saving the elegible scopes in the events, I just reverted this for now to the previous state.
I'm not sure if this presents an information leak as the information is already exposed via the `/.well-known/openid-configuration`:
```
...
"scopes_supported": [
"openid",
"roles",
"organization",
"profile",
"web-origins",
"microprofile-jwt",
"phone",
"address",
"acr",
"email",
"basic",
"offline_access"
],
...
```

Perhaps we should add a setting to the client scope configuration that allows to hide a client scope from the OIDC configuration document.

@mposolda I just tested the initial suggestion in org.keycloak.protocol.oidc.endpoints.AuthorizationEndpointChecker#checkValidScope

```
String errorMessage = "Invalid scopes: " + request.getScope();
event.detail(Details.REASON, errorMessage);
```
A scope parameter like `&scope=openid+unknown` shows up like this in the account console:
![image](https://github.com/keycloak/keycloak/assets/314690/f958fc60-579b-4e1a-83b3-4bb96126b728)

The current admin ui encodes the values shown in the events view, so I think we are fine with showing this in the table.

I also reverted the changed log levels in `OAuth2GrantTypeBase` back to their previous levels (info).

ahus1 commented:
@thomasdarimont - thank you for pointing me at the API. As it is public information already, let's consider the discussion of this outside of the scope of this issue. Sorry for the noise.

thomasdarimont committed:
Revise error message

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont committed:
Cleanup

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

mposolda reviewed (commented):
Looks good.

One thing, which I am not 100% sure, are the `info` messages logged in the `PkceUtils` class. As AFAIK the `info` level is dedicated for generic information messages related to the server rather than some error-like scenarios due the invalid request parameters etc.  I wonder if it is sufficient to remove these messages entirely as they are logged as part of the `JBossLoggingEventListener` anyway?
@thomasdarimont @ahus1 WDYT?

thomasdarimont committed:
Revised error events and logging in PkceUtils

- Added Details.REASON to events
- Removed info level logging in favor of event based logging
- Changed info level logging in isValidPkceCodeVerifier method to debug

The previously logged information in now contained in the error event.

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
@mposolda thanks for the review! I Adapted the PkceUtils to avoid info logging in favor of relying on the error events.

There were two log messages in info in `org.keycloak.protocol.oidc.utils.PkceUtils#isValidPkceCodeVerifier(..)` that I changed to debug to avoid spamming logs.

thomasdarimont committed:
Cleanup

Signed-off-by: Thomas Darimont <thomas.darimont@googlemail.com>

thomasdarimont commented:
I just looked at the same problem for the SAML Endpoints https://github.com/keycloak/keycloak/issues/29238 and noticed that the `Details.REASON` in the [SAML error events](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L217) is often a message key instead of a concrete message. Other places add a plain english error message to `Details.REASON`. 

Unfortunately the intent of `Details.REASON` is not specified explicitly... can I assume that the `Details.REASON` should always contain a human readable message?

mposolda reviewed (approved):
@thomasdarimont Thanks! IMO this is sufficient. We can possibly have a follow-up to adjust `Details.REASON` in other places if needed.

@ahus1 WDYT? Are we ok with merge this?

ahus1 reviewed (approved):
Thank you - looks great. This will help users and admin a lot, especially when they first try out Keycloak.
